### ISSUE TITLE
NPE running assets:precompile with JVM CE 22.3 and Rails 7

### ISSUE THREAD

**Issue by headius** [oracle/truffleruby#2804]
This is a simple scaffolded app but I'm not sure the app matters. It blows up during precompilation of assets.

```
rails new testapp
cd testapp
rails generate scaffold post title:string body:text published:boolean
RAILS_ENV=production rake db:migrate
RAILS_ENV=production rake assets:precompile
```

Command line:

```
Caused by: java.lang.NullPointerException: Cannot invoke "java.lang.Integer.intValue()" because "store[...]" is null
	at org.truffleruby.core.hash.library.PackedHashStoreLibrary.getHashed(PackedHashStoreLibrary.java:75)
	at org.truffleruby.core.hash.library.PackedHashStoreLibrary$LookupPackedEntryNode.getPackedArray(PackedHashStoreLibrary.java:459)
	at org.truffleruby.core.hash.library.PackedHashStoreLibraryFactory$LookupPackedEntryNodeGen.execute(PackedHashStoreLibraryFactory.java:56)
	at org.truffleruby.core.hash.library.PackedHashStoreLibrary.lookupOrDefault(PackedHashStoreLibrary.java:165)
	at org.truffleruby.core.hash.library.PackedHashStoreLibraryGen$HashStoreLibraryExports$Cached.lookupOrDefault(PackedHashStoreLibraryGen.java:239)
	at org.truffleruby.core.hash.HashNodes$GetOrUndefinedNode.getOrUndefined(HashNodes.java:223)
	at org.truffleruby.core.hash.HashNodesFactory$GetOrUndefinedNodeFactory$GetOrUndefinedNodeGen.execute(HashNodesFactory.java:781)
	at org.truffleruby.language.locals.WriteLocalVariableNode.execute(WriteLocalVariableNode.java:28)
	at org.truffleruby.language.RubyNode.doExecuteVoid(RubyNode.java:63)
	at org.truffleruby.language.control.SequenceNode.execute(SequenceNode.java:34)
	at org.truffleruby.language.RubyMethodRootNode.execute(RubyMethodRootNode.java:65)
```

**Comment by headius** [comment]
It appears to happen every time but the stack trace varies. This one has more context:

```
Cannot invoke "java.lang.Integer.intValue()" because "store[...]" is null (java.lang.NullPointerException)
	from org.truffleruby.core.hash.library.PackedHashStoreLibrary.getHashed(PackedHashStoreLibrary.java:75)
	from org.truffleruby.core.hash.library.PackedHashStoreLibrary$Set.set(PackedHashStoreLibrary.java:207)
	from org.truffleruby.core.hash.library.PackedHashStoreLibraryGen$HashStoreLibraryExports$Cached.setAndSpecialize(PackedHashStoreLibraryGen.java:142)
	from org.truffleruby.core.hash.library.PackedHashStoreLibraryGen$HashStoreLibraryExports$Cached.set(PackedHashStoreLibraryGen.java:112)
	from org.truffleruby.core.hash.library.HashStoreLibraryGen$CachedDispatch.set(HashStoreLibraryGen.java:999)
	from org.truffleruby.core.hash.library.EmptyHashStore.set(EmptyHashStore.java:49)
	from org.truffleruby.core.hash.library.EmptyHashStoreGen$HashStoreLibraryExports$Cached.setNode_AndSpecialize(EmptyHashStoreGen.java:102)
	from org.truffleruby.core.hash.library.EmptyHashStoreGen$HashStoreLibraryExports$Cached.set(EmptyHashStoreGen.java:89)
	from org.truffleruby.core.hash.HashNodes$SetIndexNode.set(HashNodes.java:239)
	from org.truffleruby.core.hash.HashNodesFactory$SetIndexNodeFactory$SetIndexNodeGen.executeAndSpecialize(HashNodesFactory.java:1043)
	from org.truffleruby.core.hash.HashNodesFactory$SetIndexNodeFactory$SetIndexNodeGen.execute(HashNodesFactory.java:991)
	from org.truffleruby.language.RubyCoreMethodRootNode.execute(RubyCoreMethodRootNode.java:48)
/Users/headius/graalvm-ce-java17-22.3.0/Contents/Home/languages/ruby/lib/gems/gems/sprockets-4.1.1/lib/sprockets/cached_environment.rb:34:in `[]='
```

This one is different but still in hash logic:

```
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 9 out of bounds for length 9
	at org.truffleruby.core.hash.library.PackedHashStoreLibrary.getHashed(PackedHashStoreLibrary.java:75)
	at org.truffleruby.core.hash.library.PackedHashStoreLibrary.promoteToBuckets(PackedHashStoreLibrary.java:134)
	at org.truffleruby.core.hash.library.PackedHashStoreLibrary$Set.set(PackedHashStoreLibrary.java:222)
	at org.truffleruby.core.hash.library.PackedHashStoreLibraryGen$HashStoreLibraryExports$Cached.set(PackedHashStoreLibraryGen.java:107)
	at org.truffleruby.core.hash.HashNodes$SetIndexNode.set(HashNodes.java:239)
```

**Comment by headius** [comment]
More context for the AIOOB:

```
Caused by:
Index 9 out of bounds for length 9 (java.lang.ArrayIndexOutOfBoundsException)
	from org.truffleruby.core.hash.library.PackedHashStoreLibrary.getHashed(PackedHashStoreLibrary.java:75)
	from org.truffleruby.core.hash.library.PackedHashStoreLibrary.promoteToBuckets(PackedHashStoreLibrary.java:134)
	from org.truffleruby.core.hash.library.PackedHashStoreLibrary$Set.set(PackedHashStoreLibrary.java:222)
	from org.truffleruby.core.hash.library.PackedHashStoreLibraryGen$HashStoreLibraryExports$Cached.setAndSpecialize(PackedHashStoreLibraryGen.java:142)
	from org.truffleruby.core.hash.library.PackedHashStoreLibraryGen$HashStoreLibraryExports$Cached.set(PackedHashStoreLibraryGen.java:112)
	from org.truffleruby.core.hash.HashNodes$SetIndexNode.set(HashNodes.java:239)
	from org.truffleruby.core.hash.HashNodesFactory$SetIndexNodeFactory$SetIndexNodeGen.execute(HashNodesFactory.java:981)
	from org.truffleruby.language.RubyCoreMethodRootNode.execute(RubyCoreMethodRootNode.java:48)
/Users/headius/graalvm-ce-java17-22.3.0/Contents/Home/languages/ruby/lib/gems/gems/sprockets-4.1.1/lib/sprockets/cache/memory_store.rb:38:in `[]='
	from /Users/headius/graalvm-ce-java17-22.3.0/Contents/Home/languages/ruby/lib/gems/gems/sprockets-4.1.1/lib/sprockets/cache/memory_store.rb:38:in `get'
	from /Users/headius/graalvm-ce-java17-22.3.0/Contents/Home/languages/ruby/lib/gems/gems/sprockets-4.1.1/lib/sprockets/cache.rb:88:in `fetch'
	from /Users/headius/graalvm-ce-java17-22.3.0/Contents/Home/languages/ruby/lib/gems/gems/sprockets-4.1.1/lib/sprockets/base.rb:71:in `file_digest'
	from /Users/headius/graalvm-ce-java17-22.3.0/Contents/Home/languages/ruby/lib/gems/gems/sprockets-4.1.1/lib/sprockets/unloaded_asset.rb:105:in `dependency_history_key'
	from /Users/headius/graalvm-ce-java17-22.3.0/Contents/Home/languages/ruby/lib/gems/gems/sprockets-4.1.1/lib/sprockets/loader.rb:324:in `fetch_asset_from_dependency_cache'
	from /Users/headius/graalvm-ce-java17-22.3.0/Contents/Home/languages/ruby/lib/gems/gems/sprockets-4.1.1/lib/sprockets/loader.rb:43:in `load'
	from /Users/headius/graalvm-ce-java17-22.3.0/Contents/Home/languages/ruby/lib/gems/gems/sprockets-4.1.1/lib/sprockets/cached_environment.rb:44:in `block in load'
```

**Comment by headius** [comment]
It finally worked after about a dozen attempts.

**Comment by eregon** [comment]
Thank you for the report. From those stacktraces it looks like that Hash is used in an unsafe concurrent way.

**Comment by eregon** [comment]
Upstream issue for this: https://github.com/rails/sprockets/issues/772

**Comment by eregon** [comment]
Fixed in sprockets 4.2.0, the same commands just work now.
